semantic_model_generator/admin_app/pages/chat.py

```diff
@@ -0,0 +1,276 @@
+    with expr_col:
+        show_expr_for_ref(message_index)
+    with sql_edit_col:
+        sql_without_cte = remove_ltable_cte(sql)
+        user_updated_sql = st.text_area(
+            "Edit SQL below (write SQL against logical tables/columns on the left):",
+            value=sql_without_cte,
+            height=250,
+            help="This is the original SQL as generated by the Model. Please update as needed to create a sql suitable for your verified query repository.",
+            key=f"updated_sql_key_{message_index}",
+        )
+        st.caption("👆 You can edit the SQL query above to correct or optimize it.")
+        confirm_button = st.button(
+            "Click to confirm edits and compile",
+            key=f"confirm_edits_idx_{message_index}",
+        )
+        if confirm_button:
+            st.session_state.confirmed_edits = True
+
+        if st.session_state.confirmed_edits:
+            try:
+                sql_to_execute = expand_all_logical_tables_as_ctes(
+                    user_updated_sql, st.session_state.ctx
+                )
+                df = pd.read_sql(sql_to_execute, conn)
+                st.code(user_updated_sql)
+                st.dataframe(df)
+                if st.button(
+                    "Click to save modified query to yaml",
+                    key=f"confirm_vqr_idx_{message_index}",
+                ):
+                    add_verified_query(question, sql)
+                    st.session_state.editing = False
+                    st.session_state.confirmed_edits = False
+            except Exception as e:
+                raise ValueError(
+                    f"Edited SQL not compatible with semantic model provided, please double check: {e}"
+                )
+
+
+def add_verified_query(question: str, sql: str) -> None:
+    """Save verified question and SQL into an in-memory list with additional details."""
+    # Verified queries follow the Snowflake definitions.
+    verified_query = semantic_model_pb2.VerifiedQuery(
+        question=question,
+        sql=sql,
+        verified_by=USER,
+        verified_at=int(time.time()),
+    )
+    st.session_state.semantic_model.verified_queries.append(verified_query)
+    st.success(
+        "Verified Query Added! You can go back to validate your YAML again and upload; or keep adding more verified queries."
+    )
+
+
+def display_content(
+    conn: SnowflakeConnection,
+    content: list[dict[str, Any]],
+    message_index: int | None = None,
+) -> None:
+    """Displays a content item for a message. For generated SQL, allow user to add to verified queries directly or edit then add."""
+    message_index = message_index or len(st.session_state.messages)
+    sql = ""
+    question = ""
+    for item in content:
+        if item["type"] == "text":
+            if question == "" and "__" in item["text"]:
+                question = item["text"].split("__")[1]
+            if "<SUGGESTION>" in item["text"]:
```

Comment:
nit: add pointer to where this constant is defined

Reply:
added a comment